src/core/lib/resource_quota/api.cc

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/resource_quota/api.h"
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/gpr/useful.h"
+
+namespace grpc_core {
+
+ResourceQuotaPtr ResourceQuotaFromChannelArgs(const grpc_channel_args* args) {
+  return grpc_channel_args_find_pointer<ResourceQuota>(args,
+                                                       GRPC_ARG_RESOURCE_QUOTA)
+      ->Ref();
+}
+
+namespace {
+grpc_arg MakeArg(ResourceQuota* quota) {
+  grpc_arg arg;
+  arg.type = GRPC_ARG_POINTER;
+  arg.key = const_cast<char*>(GRPC_ARG_RESOURCE_QUOTA);
+  arg.value.pointer.p = quota;
+  arg.value.pointer.vtable = grpc_resource_quota_arg_vtable();
+  return arg;
+}
+}  // namespace
+
+grpc_channel_args* EnsureResourceQuotaInChannelArgs(
+    const grpc_channel_args* args) {
+  if (grpc_channel_args_find(args, GRPC_ARG_RESOURCE_QUOTA) != nullptr) {
+    return grpc_channel_args_copy(args);
+  }
+  // If there's no existing quota, add it to the default one - shared between
+  // all channel args declared thusly. This prevents us from accidentally not
+  // sharing subchannels due to their channel args not specifying a quota.
+  static auto default_resource_quota = MakeResourceQuota();
+  auto new_arg = MakeArg(default_resource_quota->Ref().release());
+  return grpc_channel_args_copy_and_add(args, &new_arg, 1);
+}
+
+grpc_channel_args* ChannelArgsWrappingResourceQuota(
+    ResourceQuotaPtr resource_quota) {
+  auto new_arg = MakeArg(resource_quota.release());
+  return grpc_channel_args_copy_and_add(nullptr, &new_arg, 1);
+}
+
+}  // namespace grpc_core
+
+extern "C" const grpc_arg_pointer_vtable* grpc_resource_quota_arg_vtable() {
+  static const grpc_arg_pointer_vtable vtable = {
+      // copy
+      [](void* p) -> void* {
+        return static_cast<grpc_core::ResourceQuota*>(p)->Ref().release();
+      },
+      // destroy
+      [](void* p) { static_cast<grpc_core::ResourceQuota*>(p)->Unref(); },
+      // compare
+      [](void* p, void* q) { return grpc_core::QsortCompare(p, q); }};
+  return &vtable;
+}
+
+extern "C" grpc_resource_quota* grpc_resource_quota_create(const char*) {
+  return reinterpret_cast<grpc_resource_quota*>(new grpc_core::ResourceQuota());
```

Comment:
Interesting. The `grpc_resource_quota` type becomes meaningless in this API, yeah? The type itself has no bearing on what we do with that pointer anymore. If I understand that correctly, I'm not sure if we should be using the C API this way.

Reply:
So everywhere it becomes some pointer to a thing unknown, and in well defined places we cast it to a thing that has a definition.